graders/java/JUnitAutograder.java

```diff
@@ -0,0 +1,242 @@
+
+        LauncherDiscoveryRequestBuilder requestBuilder = LauncherDiscoveryRequestBuilder.request();
+
+        for (String className : testClasses) {
+            try {
+                Class<?> cls = Class.forName(className, true, Thread.currentThread().getContextClassLoader());
+                LauncherDiscoveryRequest request = LauncherDiscoveryRequestBuilder.request().selectors(selectClass(cls)).build();
+                this.launcher.execute(request);
+            } catch (ClassNotFoundException e) {
+                this.points = 0;
+                this.message = "Could not load test class " + className;
+                throw new UngradableException();
+            }
+        }
+    }
+
+    @Override
+    public synchronized void executionFinished(TestIdentifier test, TestExecutionResult result) {
+        if (!test.isTest()) return;
+        AutograderTest autograderTest = new AutograderTest(test.getDisplayName());
+
+        for (TestTag tag : test.getTags()) {
+            if (tag.getName().startsWith("points=")) {
+                try {
+                    autograderTest.maxPoints = autograderTest.points = Double.parseDouble(tag.getName().substring(7));
+                } catch(NumberFormatException exception) {
+                    this.output = "Could not parse points tag: " + tag.getName();
+                    this.gradable = false;
+                }
+            }
+        }
+
+        // For compatibility with JUnit4 autograder
+        test.getSource().ifPresent(source -> {
+            if (source instanceof MethodSource) {
+                AutograderInfo info = ((MethodSource) source).getJavaMethod().getAnnotation(AutograderInfo.class);
+                if (info != null) {
+                    if (info.points() > 0)
+                        autograderTest.points = autograderTest.maxPoints = info.points();
+                    if (!"".equals(info.name()))
+                        autograderTest.name = info.name();
+                    if (!"".equals(info.description()))
+                        autograderTest.description = info.description();
+                }
+            }
+        });
+
+        if (!result.getStatus().equals(TestExecutionResult.Status.SUCCESSFUL)) {
+            autograderTest.points = 0;
+            result.getThrowable().ifPresent(t -> autograderTest.message = t.getMessage());
+            if (autograderTest.message == null)
+                autograderTest.message = "";
+        }
+
+        this.tests.add(autograderTest);
+        this.points += autograderTest.points;
+        this.maxPoints += autograderTest.maxPoints;
+    }
+
+    private void saveResults() {
+
+        JSONArray resultsTests = new JSONArray();
+        //Collections.sort(this.tests);
+        for (AutograderTest test : this.tests)
+            resultsTests.add(test.toJson());
+
+        JSONObject results = new JSONObject();
+        results.put("score", this.maxPoints > 0 ? this.points / this.maxPoints : 0);
```

Comment:
I can't figure out why this object doesn't include `succeeded` which the spec for external grading says is required (https://prairielearn.readthedocs.io/en/stable/externalGrading/)? However when I try to debug locally it seems that something in between is adding the `succeeded` field. But if I am reading the docs right the `succeeded` field should be added here.

Reply:
Seems like that documentation is incorrect. If I understand the code correctly (https://github.com/PrairieLearn/PrairieLearn/blob/ba42f43cc91adcef1f883a1ebad5ab927923aeeb/grader_host/index.js#L514), `succeeded` is ignored and set to true as long as the results.json file is properly parsed. @mwest1066 can you confirm this is indeed the case? In any case, I just `gradable` for what is described in the docs as `succeeded`, and that works for the C grader as well.

Reply:
I think this is actually a bug in `grader_host`. We are trying to distinguish between three different types of error:
1. Error in the grader code itself. This is indicated by `succeeded = false`. Ideally this should never happen. In this case we will ignore any other data returned by the grader. This grading attempt won't count against the student in any way (no points lost, no attempts used).
2. Error in the student code for which we want to abandon grading. This is indicated by setting `succeeded = true` and `gradable = false`. This will be reported back to the student as an "Invalid" submission (hopefully with detailed information from the grader about what was wrong). This grading attempt won't count against the student in any way (no points lost, no attempts used).
3. Test case errors in the student code. These are "normal" errors which cause test cases to fail. We return `succeeded = true` and `gradable = true` and points are calculated based on the number of test cases that pass.

I think that `grader_host` should not be setting `succeeded = true`. Instead, I think that it should be doing something like:
```
results.succeeded = (results.succeeded == null ? true : results.succeeded);
```
That is, if `succeeded` is missing then we'll set it to `true` to handle old graders that aren't setting it correctly.

Ideally we would do `results.succeeded = !!results.succeeded`, which would force graders to return `succeeded = true` if they want results to be processed. But I worry that this would break backwards compatibility.

@jonatanschroeder does this sound correct to you? Should we change `grader_host` and or the docs?

Reply:
It seems the docs @falkirks was referring to are not the latest version. The latest version (https://prairielearn.readthedocs.io/en/latest/externalGrading/) doesn't make any reference at all to `succeeded`. @mwest1066  how do you want to proceed?

Reply:
Good catch! Let's leave all this as is. No changes needed.